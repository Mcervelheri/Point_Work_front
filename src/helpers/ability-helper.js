import { Ability } from '@casl/ability';
import { Can as AbilityCan } from '@casl/react';
import { connect } from 'react-redux';

import store from '../redux/store';

const mapStateToProps = ({ ability }) => ({ ability });

export const Can = connect(mapStateToProps)(AbilityCan);

const detectSubjectType = subject => {
    if (subject?.modelName) {
        return subject.modelName;
    }
    if (subject?.constructor) {
        return subject.constructor.modelName || subject.constructor.name;
    }
    return subject;
};

export function createAbility(rules = []) {
    return new Ability(rules, {
        detectSubjectType,
    });
}

/**
 * Renderiza o componente em "children" caso o usuário tenha permissão
 * senão renderiza o componente em "defaultChildren" que tem valor padrão "null" se não informado.
 * @param {string|string[]} actions Ação para validação. Se informado um array, a condição é aplicada como "ou".
 * @param {string} subject Modelo para o qual a ação é aplicada
 * @param {object} children JSX para renderizar em caso de permissão existente
 * @param {object} defaultChildren JSX para renderizar em caso de permissão não existente
 */
export function renderIfICan(actions, subject, children, defaultChildren = null) {
    const { ability } = store.getState();
    const shouldShow = Array.isArray(actions)
        ? actions.some(action => ability.can(action, subject))
        : ability.can(actions, subject);

    return shouldShow ? children : defaultChildren;
}
