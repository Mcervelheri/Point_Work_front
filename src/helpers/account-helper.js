import React from 'react';

import { InfoCircleOutlined } from '@ant-design/icons';
import { Tooltip, Tag } from 'antd';

import { PREFIX_ACCOUNT_IN_DOCK_HOMOLOG, PREFIX_ACCOUNT_IN_DOCK_PRODUCTION } from '../values/constants';
import {
    STATUS_TYPES,
    SPD_STATUS,
    STATE_OPTIONS,
    COMPANY_FORMAT,
    GENDER_OPTIONS,
    MARITAL_OPTIONS,
    PROFESSION_OPTIONS,
    ACCOUNT_PROFILE_OPTIONS,
    TYPE_PHONE_OPTIONS,
    STATUS_TYPES_RECHARGES,
} from '../values/enums';
import { PARTNERS_TYPE_OPTIONS } from '../values/enums/accounts';

import { maskWithVerificationDigit } from './masks';

const { REACT_APP_ENV } = process.env;

export const getAccountStatusConfig = statusKey => {
    return STATUS_TYPES.find(statusItem => statusItem.key === statusKey);
};

export const getAccountSPDConfig = spdKey => {
    const startsWithZero = spdKey.startsWith('0');

    let spd = spdKey;

    if (!startsWithZero) {
        spd = `0${spdKey}`;
    }

    return SPD_STATUS.find(spdItem => spdItem.key === spd);
};

export const getClientCompany = arr => {
    if (!arr) return null;
    return arr.find(item => item.accountProfile === 'MAIN');
};

export const getClientPersonArr = arr => {
    if (!arr) return null;
    return arr.filter(item => item.accountProfile !== 'MAIN');
};

export const getClientById = (arr, id) => {
    if (!arr) return null;
    return arr.find(item => `${item.idClient}` === id);
};

export const convertMainPhoneToString = mainPhone => {
    if (mainPhone?.areaCode && mainPhone?.number) {
        const { areaCode, number } = mainPhone;

        return `${parseInt(areaCode, 10)}${number}`;
    }
    return null;
};

export const transformStringToMainPhone = phoneNumber => {

    const areaCode = `0${phoneNumber.substr(0, 2)}`;
    const number = phoneNumber.substr(2);
    const countryCode = '55';

    return { areaCode, number, countryCode };
};

export const getAccountNumber = externalId => {

    if (externalId && externalId !== 'null') {
        return externalId;
    }

    return null;
};

export const extractOnboardingExtraData = result => {
    const { data } = result;

    if (!data || data.length <= 0) return null;

    const company = data[0];
    const idRegistration = getClientCompany(company.clients)?.idRegistration;

    return {
        openAt: company?.openAt,
        externalId: company?.externalId,
        status: company?.status,
        idRegistration,
    };

};

export const getStateOption = stateKey => {
    if (!stateKey) return null;
    return STATE_OPTIONS.find(option => `${option.key}` === `${stateKey}`);
};

export const getEstablishmentFormatOption = eFormatKey => {
    if (!eFormatKey) return null;
    return COMPANY_FORMAT.find(option => `${option?.key}` === `${eFormatKey}`);
};

export const getGenderOption = genderKey => {
    if (!genderKey) return null;
    return GENDER_OPTIONS.find(option => `${option?.key}` === `${genderKey}`);
};

export const getMaritalOption = maritalKey => {
    if (!maritalKey) return null;
    return MARITAL_OPTIONS.find(option => `${option?.key}` === `${maritalKey}`);
};

export const getProfessionalOption = professionalKey => {
    if (!professionalKey) return null;
    return PROFESSION_OPTIONS.find(option => `${option?.key}` === `${professionalKey}`);
};

export const getAccProfileOptions = accProfileKey => {
    if (!accProfileKey) return null;
    return ACCOUNT_PROFILE_OPTIONS.find(option => `${option?.key}` === `${accProfileKey}`);
};

export const getPhoneType = phoneKey => {
    if (!phoneKey) return null;
    return TYPE_PHONE_OPTIONS.find(option => `${option?.key}` === `${phoneKey}`);
};

export const getPartnerTypeOption = (accProfile, type) => {
    if (!accProfile || !type) return PARTNERS_TYPE_OPTIONS[0];
    return PARTNERS_TYPE_OPTIONS.find(data => data.profile === accProfile && data.type === type);
};

export const getRechargeStatusConfig = statusKey => {
    return STATUS_TYPES_RECHARGES.find(statusItem => statusItem.key === statusKey);
};

export const getAccountWithMask = userAccount => {
    const accountPrefix = REACT_APP_ENV === 'production'
        ? PREFIX_ACCOUNT_IN_DOCK_PRODUCTION
        : PREFIX_ACCOUNT_IN_DOCK_HOMOLOG;

    return maskWithVerificationDigit(`${accountPrefix}${userAccount}`);
};

export const mapSPD = spdList => {
    if (!spdList) return null;
    return spdList.map(spd => {
        const config = getAccountSPDConfig(spd);
        return (
            <Tooltip
                placement="bottom"
                title={config?.label}
                key={spd}
            >
                <Tag color="default">
                    <span>
                        {config?.key}
                    </span>
                    <InfoCircleOutlined />
                </Tag>
            </Tooltip>
        );
    });
};
