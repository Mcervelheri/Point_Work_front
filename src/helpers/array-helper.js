export function removeItem(array, item) {
    const newArray = [...array];
    for (let i = 0; i < newArray.length; i += 1) {
        if (newArray[i] === item) {
            newArray.splice(newArray.indexOf(item), 1);
            i -= 1;
        }
    }

    return newArray;
}

export const removeAllNullablesItens = array => {
    return removeItem(array, null);
};
