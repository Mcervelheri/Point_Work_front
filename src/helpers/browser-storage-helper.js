
export default {
    setItem(key, value) {
        if (window && window.localStorage) {
            localStorage.setItem(key, value);
        }
    },

    getItem(key) {
        if (!window || !window.localStorage) {
            return {};
        }

        return localStorage.getItem(key);
    },

    saveColumnsConfig(pageId, columns) {
        this.setItem(`${pageId}-columns`, JSON.stringify(columns));
    },

    loadColumnsConfig(pageId) {
        const columns = this.getItem(`${pageId}-columns`);
        return columns ? JSON.parse(columns) : null;
    },

    saveFilters(pageId, filters) {
        if (window && window.localStorage) {
            localStorage.setItem(`${pageId}-filters`, JSON.stringify(filters));
        }
    },

    loadFilters(pageId) {
        if (!window || !window.localStorage) {
            return {};
        }

        try {
            const content = localStorage.getItem(`${pageId}-filters`);
            return content ? JSON.parse(content) : {};
        } catch (err) {
            return {};
        }
    },
};
