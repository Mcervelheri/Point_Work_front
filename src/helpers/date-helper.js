
import moment from 'moment-timezone';

export const disabledDate = (value, dates) => {
    const futureDate = value && value > moment().endOf('day');

    if (!dates || dates.length === 0) {
        return futureDate;
    }

    const tooLate = dates[0] && value.diff(dates[0], 'days') > 30;
    const tooEarly = dates[1] && dates[1].diff(value, 'days') > 30;

    return tooLate || futureDate || tooEarly;
};

/**
 * Faz a formatação da data fornecida e retorna no formato especificado
 * @param {string} date - Data que deverá ser formatada
 * @param {string} format - Padrão de data que será usado na formatação
 * @param {string} inputFormat - Padrão da data de entrada para datas diferentes do padrão ISO 8601
 * @returns {(string|null)} String da data formatada ou null caso não tenha sido informado uma data
 */
export const formatDate = (date, format, inputFormat) => {
    if (date) {
        return moment(date, inputFormat).format(format);
    }

    return null;
};

/**
 * Faz a formatação da data atual e retorna no formato especificado
 * @param {string} format - Padrão de data que será usado na formatação
 * @returns {string} String da data atual no formato especificado
 */
export const formatCurrentDate = format => {
    return moment().format(format);
};

/**
 * Faz a validação se a data informada é uma data válida
 * @param {string} date - Data de referência para validação
 * @param {string} datePattern - Padrão da data de referência
 * @returns {boolean} true caso a data seja considerada válida e false para o contrário
 * @see {@link https://momentjs.com/docs/#/parsing/is-valid/}
 */
export const isValidDate = (date, datePattern) => {
    return moment(date, datePattern).isValid();
};

/**
 * Faz a subtração de uma quantidade de dias a partir da data atual
 * @param {number} amount - Quantidade de dias para subtrair
 * @param {string} [format=YYYY-MM-DD] - Padrão de data que será usado na formatação
 * @returns {string} String da data resultante no formato YYYY-MM-DD (padrão)
 * @see {@link https://momentjs.com/docs/#/manipulating/subtract/}
 */
export const subtractDaysFromToday = (amount, format = 'YYYY-MM-DD') => {
    return moment().subtract(amount, 'days').format(format);
};
