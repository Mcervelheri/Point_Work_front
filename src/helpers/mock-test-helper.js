import React from 'react';

import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import * as redux from 'react-redux';
import { Provider } from 'react-redux';
import { Router, MemoryRouter } from 'react-router-dom';
import { create } from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

const defaultConfig = {
    useSelectorValues: {},
    mockDispatch: false,
    mockLocation: false,
    locationState: {},
    pathName: '/',
    search: '',
};

const defineUsedMocks = (config, children) => {

    let component = children;
    const mockLocationInitialEntries = [];

    if (config.useSelectorValues) {
        const spySelector = jest.spyOn(redux, 'useSelector');
        spySelector.mockReturnValue(config.useSelectorValues);
    }

    if (config.mockDispatch) {
        const spyDispatch = jest.spyOn(redux, 'useDispatch');
        const mockDispatchFn = jest.fn();
        spyDispatch.mockReturnValue(mockDispatchFn);
    }

    if (config.mockLocation) {

        const mockLocationObject = {
            pathname: config.pathName,
            search: config.search,
            hash: '',
            state: config.locationState,
            key: '5nvxpbdafa',
        };

        mockLocationInitialEntries.push(mockLocationObject);

        component = (
            <MemoryRouter initialEntries={mockLocationInitialEntries}>
                {children}
            </MemoryRouter>
        );
    }

    const history = createMemoryHistory({
        initialEntries: [''],
    });
    if (config.locationState) {
        const state = { ...config.locationState };
        history.push(config.pathName, state);
    }
    return { history, component };
};
export const renderWithMocks = (children, config = defaultConfig) => {
    const { history, component } = defineUsedMocks(config, children);
    const mockStore = configureMockStore();
    const store = mockStore({
        ...config.useSelectorValues,
    });

    return (
        render(
            <Router history={history}>
                <Provider store={store}>
                    {component}
                </Provider>
            </Router>,
        )
    );
};
export const createWithMocks = (children, config = defaultConfig) => {
    const { history, component } = defineUsedMocks(config, children);
    const mockStore = configureMockStore();
    const store = mockStore({
        ...config.useSelectorValues,
    });

    return (
        create(
            <Router history={history}>
                <Provider store={store}>
                    {component}
                </Provider>
            </Router>,
        )
    );
};
