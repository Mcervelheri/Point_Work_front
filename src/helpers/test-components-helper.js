import React, { useEffect } from 'react';

import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';

import thunks from '../redux/thunks';

export const ElementUseSelector = () => {

    const mockValues = useSelector(state => state.values);

    return (
        <span>{mockValues.title}</span>
    );
};

export const ElementUseDispatch = () => {

    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(thunks.usuario.refreshUserAuthentication());
    }, [dispatch]);

    return (
        <span>use dispatch element</span>
    );
};

export const ElementUseLocation = () => {
    const location = useLocation();

    return (
        <div>
            <span>Location props:</span>
            <span>State example: {location.state?.title}</span>
            <span>key: {location.key}</span>
            <span>pathname: {location.pathname}</span>
            <span>search: {location.search}</span>
        </div>
    );
};
