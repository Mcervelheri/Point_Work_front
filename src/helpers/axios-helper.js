import { messageError } from '../helpers/toast';
import store from '../redux/store';
import thunks from '../redux/thunks';

export function tokenInterceptor(request) {
    const { token } = store.getState();
    if (token && !request?.headers?.token) {
        request.headers.Authorization = `Bearer ${token}`;
    }
    // request.headers.timezone = moment.tz.guess(); // Timezone
    return request;
}

export function logout401(responseError) {

    const { response } = responseError;

    if (response?.status === 401 && store?.getState()?.token) {
        store.dispatch(thunks.usuario.removeAuthenticationData());
        messageError('Token expirado. Fa√ßa o login novamente.');
        return responseError;
    }

    return Promise.reject(responseError);
}
