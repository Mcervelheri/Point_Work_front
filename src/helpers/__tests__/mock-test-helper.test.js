import React from 'react';

import { screen } from '@testing-library/react';

import { createWithMocks, renderWithMocks } from '../helpers/mock-test-helper';
import { ElementUseDispatch, ElementUseLocation, ElementUseSelector } from '../helpers/test-components-helper';

describe('Quando usada a função renderWithMocks', () => {

    describe('e o componente utiliza os hooks do redux', () => {
        beforeEach(() => {
            jest.clearAllMocks();
        });

        it('retorna o valor passado pela config do useSelectorValues', () => {

            const defaultConfig = {
                useSelectorValues: {
                    title: 'mocked title',
                },
            };

            renderWithMocks(<ElementUseSelector />, defaultConfig);

            expect(screen.getByText('mocked title')).toBeInTheDocument();
        });

        it(`retorna um erro quando um componente que possui um useDispatch no seu useEffect não recebe a config
        mockDispatch como true`, () => {

            try {
                renderWithMocks(<ElementUseDispatch />, {});
            } catch (error) {
                expect(error).toBeInstanceOf(Error);
            }

        });

        it(`renderiza corretamente um componente que possui um useDispatch no seu useEffect 
        quando passada a config mockDispatch true`, () => {

            const defaultConfig = {
                mockDispatch: true,
            };

            renderWithMocks(<ElementUseDispatch />, defaultConfig);

            expect(screen.getByText('use dispatch element')).toBeInTheDocument();
        });
    });

    describe('e o componente utiliza o hook useLocation', () => {

        it('mocka o useLocation quando o atributo mockLocation da config for true', () => {

            const defaultConfig = {
                mockLocation: true,
            };

            renderWithMocks(<ElementUseLocation />, defaultConfig);

            expect(screen.getByText('key: 5nvxpbdafa')).toBeInTheDocument();
        });

        it('não mocka o useLocation quando o atributo mockLocation da config for false', () => {

            const defaultConfig = {
                mockLocation: false,
            };

            renderWithMocks(<ElementUseLocation />, defaultConfig);

            expect(screen.queryByText('key: 5nvxpbdafa')).not.toBeInTheDocument();
        });

        it('retorna o atributo title passado no locationState', () => {

            const defaultConfig = {
                mockLocation: true,
                locationState: {
                    title: 'use location test',
                },
            };

            renderWithMocks(<ElementUseLocation />, defaultConfig);

            expect(screen.getByText('State example: use location test')).toBeInTheDocument();
        });

        it('retorna o path mockado através do atributo pathName do config', () => {

            const defaultConfig = {
                mockLocation: true,
                pathName: '/test-location-state',
            };

            renderWithMocks(<ElementUseLocation />, defaultConfig);

            expect(screen.getByText('pathname: /test-location-state')).toBeInTheDocument();
        });

        it('retorna o search mockado através do atributo search do config', () => {

            const defaultConfig = {
                mockLocation: true,
                search: '?limit=10&page=1&sort=desc',
            };

            renderWithMocks(<ElementUseLocation />, defaultConfig);

            expect(screen.getByText('search: ?limit=10&page=1&sort=desc')).toBeInTheDocument();
        });
    });
});

describe('Quando usada a função createWithMocks', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renderiza corretamente o valor passado pela config do useSelectorValues', () => {

        const defaultConfig = {
            useSelectorValues: {
                title: 'mocked title',
            },
        };

        const element = createWithMocks(<ElementUseSelector />, defaultConfig);

        expect(element.toJSON()).toMatchInlineSnapshot(`
           <span>
             mocked title
           </span>
        `);
    });

    it('renderiza corretamente os valores passados na config relacionados ao mock do useLocation', () => {

        const defaultConfig = {
            mockLocation: true,
            locationState: {
                title: 'use location test',
            },
            pathName: '/test-location-state',
            search: '?limit=10&page=1&sort=desc',
        };

        const element = createWithMocks(<ElementUseLocation />, defaultConfig);

        expect(element.toJSON()).toMatchSnapshot();
    });
});
