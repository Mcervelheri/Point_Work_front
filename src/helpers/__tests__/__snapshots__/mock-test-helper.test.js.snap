// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quando usada a função createWithMocks renderiza corretamente os valores passados na config relacionados ao mock do useLocation 1`] = `
<div>
  <span>
    Location props:
  </span>
  <span>
    State example: 
    use location test
  </span>
  <span>
    key: 
    5nvxpbdafa
  </span>
  <span>
    pathname: 
    /test-location-state
  </span>
  <span>
    search: 
    ?limit=10&page=1&sort=desc
  </span>
</div>
`;
